### PR TITLE
fix: use server-side-apply for all operations

### PR DESCRIPTION
**Pull request by kadel**
## Description
This changes all Patch and Create operations to use server-side apply.
This should lower risks of conflicts in  situations where there is another operator that modifies resources or when admission controllers changes resources.

## Which issue(s) does this PR fix or relate to

https://issues.redhat.com/browse/RHIDP-4189

## PR acceptance criteria

- [x] Tests
- [ ] Documentation
- [ ] If the bundle manifests have been updated, make sure to review the [`rhdh-operator.csv.yaml`](../.rhdh/bundle/manifests/rhdh-operator.csv.yaml) file accordingly

### CODE REVIEW

**Review by gazarenkov — APPROVED** [review]
Like this approach LGTM, let's observe and see if we can even more optimize it.

File: controllers/backstage_controller.go
Comment (rm3l) [inline controllers/backstage_controller.go:245]: nit: Can we remove all these commented lines?

**Review by rm3l — COMMENTED** [review]
Thanks for this!
I tested it, and while it works great for new installs of the operator, I came across an issue when testing the upgrade path, from 1.2.x to `main`, with an existing operator-backed deployment.

Steps to reproduce the upgrade path (which can also be reproduced by running `make test-e2e-upgrade [IMG=<image-build-of-the-operator-from-this-branch>]`):
1. Switch to the 1.2.x branch and run the operator with `make install run`
2. In a separate terminal, create a CR. I used the simple https://github.com/redhat-developer/rhdh-operator/blob/1.2.x/examples/bs1.yaml
3. Wait for the application pods to be running and ready
4. Stop the operator running from step 1
5. Switch to this PR branch and run the operator with `make install run`

At this point, it should reconcile successfully, but it looks like there is a reconciliation error:
```
$ oc describe backstage bs1                                  
Name:         bs1
Namespace:    my-ns
Labels:       <none>
Annotations:  <none>
API Version:  rhdh.redhat.com/v1alpha2
Kind:         Backstage
Metadata:
  Creation Timestamp:  2024-08-21T09:14:08Z
  Generation:          1
  Resource Version:    65153
  UID:                 eb0a158e-01a6-4b9e-84d3-8fc61fc43072
Status:
  Conditions:
    Last Transition Time:  2024-08-21T09:19:41Z
    Message:               failed to apply backstage objects failed to create secret: Apply failed with 1 conflict: conflict with "manager" using v1: .metadata.ownerReferences[uid="eb0a158e-01a6-4b9e-84d3-8fc61fc43072"]
    Reason:                DeployFailed
    Status:                False
    Type:                  Deployed
Events:                    <none>
```

I found a lot of these errors in the operator logs:
```
failed to apply backstage objects failed to create secret: Apply failed with 1 conflict: conflict with \"manager\" using v1: .metadata.ownerReferences[uid=\"eb0a158e-01a6-4b9e-84d3-8fc61fc43072\"]
```

<details>
<summary>Full logs</summary>

```
☸ my-ns/api-ci-ln-5nxf3cb-72292-origin-ci-int-gce-dev-rhcloud-com:6443/kube:admin (my-ns) in ~/w/p/b/j/operator on  use-server-side-apply [$!] via 🐹 v1.21.12
$ make install run                                                                                                                                                         11:19:28 [160/10442]
test -s /home/asoro/work/projects/backstage/janus-idp/operator/bin/controller-gen || GOBIN=/home/asoro/work/projects/backstage/janus-idp/operator/bin go install sigs.k8s.io/controller-tools/c
md/controller-gen@v0.14.0                                                                                                                                                                      
/home/asoro/work/projects/backstage/janus-idp/operator/bin/controller-gen rbac:roleName=manager-role crd webhook paths="./..." output:crd:artifacts:config=config/crd/bases                    
test -s /home/asoro/work/projects/backstage/janus-idp/operator/bin/kustomize || { curl -Ss "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh" | bas
h -s -- 5.4.2 /home/asoro/work/projects/backstage/janus-idp/operator/bin; }                                                                                                                    
/home/asoro/work/projects/backstage/janus-idp/operator/bin/kustomize build config/crd | kubectl apply -f -                                                                                     
customresourcedefinition.apiextensions.k8s.io/backstages.rhdh.redhat.com configured                                                                                                            
/home/asoro/work/projects/backstage/janus-idp/operator/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."                                                            
test -s /home/asoro/work/projects/backstage/janus-idp/operator/bin/goimports || GOBIN=/home/asoro/work/projects/backstage/janus-idp/operator/bin go install golang.org/x/tools/cmd/goimports@v0
.16.1                                                                                                                                                                                          
find . -not -path '*/\.*' -name '*.go' -exec /home/asoro/work/projects/backstage/janus-idp/operator/bin/goimports -w {} \;                                                                     
go vet ./...                                                                                                                                                                                   
go build -o bin/manager main.go                                                                                                                                                                
cd /home/asoro/work/projects/backstage/janus-idp/operator/bin && mkdir -p default-config && cp ../config/manager/default-config/* default-config && ./manager
2024-08-21T11:19:40+02:00       INFO    setup   starting manager with parameters:       {"own-runtime": true, "env.LOCALBIN": "/home/asoro/work/projects/backstage/janus-idp/operator/bin", "is
OpenShift": true}
2024-08-21T11:19:40+02:00       INFO    controller-runtime.metrics      Starting metrics server
2024-08-21T11:19:40+02:00       INFO    starting server {"kind": "health probe", "addr": "[::]:8081"}
2024-08-21T11:19:40+02:00       INFO    controller-runtime.metrics      Serving metrics server  {"bindAddress": ":8080", "secure": false}
2024-08-21T11:19:40+02:00       INFO    Starting EventSource    {"controller": "backstage", "controllerGroup": "rhdh.redhat.com", "controllerKind": "Backstage", "source": "kind source: *v1alp
ha2.Backstage"}
2024-08-21T11:19:40+02:00       INFO    Starting EventSource    {"controller": "backstage", "controllerGroup": "rhdh.redhat.com", "controllerKind": "Backstage", "source": "kind source: *v1.Pa
rtialObjectMetadata"}
2024-08-21T11:19:40+02:00       INFO    Starting EventSource    {"controller": "backstage", "controllerGroup": "rhdh.redhat.com", "controllerKind": "Backstage", "source": "kind source: *v1.Pa
rtialObjectMetadata"}
2024-08-21T11:19:40+02:00       INFO    Starting Controller     {"controller": "backstage", "controllerGroup": "rhdh.redhat.com", "controllerKind": "Backstage"}
2024-08-21T11:19:40+02:00       INFO    Starting workers        {"controller": "backstage", "controllerGroup": "rhdh.redhat.com", "controllerKind": "Backstage", "worker count": 1}
2024-08-21T11:19:41+02:00       DEBUG   failed to patch object => trying to delete it (and losing any custom labels/annotations on it) so it can be recreated upon next reconciliation...     {
"controller": "backstage", "controllerGroup": "rhdh.redhat.com", "controllerKind": "Backstage", "Backstage": {"name":"bs1","namespace":"my-ns"}, "namespace": "my-ns", "name": "bs1", "reconcil
eID": "e2fbadf5-279c-4024-bb1b-57c8c631e709", "*v1.ConfigMap": "backstage-appconfig-bs1", "cause": "failed to patch object *v1.ConfigMap: Apply failed with 2 conflicts: conflicts with \"manag
er\" using v1:\n- .data.default.app-config.yaml\n- .metadata.ownerReferences[uid=\"eb0a158e-01a6-4b9e-84d3-8fc61fc43072\"]"}
2024-08-21T11:19:41+02:00       DEBUG   deleted object. If you had set any custom labels/annotations on it manually, you will need to add them again    {"controller": "backstage", "controller
Group": "rhdh.redhat.com", "controllerKind": "Backstage", "Backstage": {"name":"bs1","namespace":"my-ns"}, "namespace": "my-ns", "name": "bs1", "reconcileID": "e2fbadf5-279c-4024-bb1b-57c8c63
1e709", "*v1.ConfigMap": "backstage-appconfig-bs1"}
2024-08-21T11:19:41+02:00       ERROR   Reconciler error        {"controller": "backstage", "controllerGroup": "rhdh.redhat.com", "controllerKind": "Backstage", "Backstage11:19:42 [125/10442]
mespace":"my-ns"}, "namespace": "my-ns", "name": "bs1", "reconcileID": "e2fbadf5-279c-4024-bb1b-57c8c631e709", "error": "failed to apply backstage objects failed to create secret: Apply faile
d with 1 conflict: conflict with \"manager\" using v1: .metadata.ownerReferences[uid=\"eb0a158e-01a6-4b9e-84d3-8fc61fc43072\"]"}                                                               
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler                                                                                                          
        /home/asoro/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:329
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /home/asoro/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /home/asoro/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:227
2024-08-21T11:19:42+02:00       DEBUG   create object   {"controller": "backstage", "controllerGroup": "rhdh.redhat.com", "controllerKind": "Backstage", "Backstage": {"name":"bs1","namespace"
:"my-ns"}, "namespace": "my-ns", "name": "bs1", "reconcileID": "b28cbcbf-8ae1-4072-9e2a-d235fed2c014", "*v1.ConfigMap": "backstage-appconfig-bs1"}
2024-08-21T11:19:42+02:00       ERROR   Reconciler error        {"controller": "backstage", "controllerGroup": "rhdh.redhat.com", "controllerKind": "Backstage", "Backstage": {"name":"bs1","na
mespace":"my-ns"}, "namespace": "my-ns", "name": "bs1", "reconcileID": "b28cbcbf-8ae1-4072-9e2a-d235fed2c014", "error": "failed to apply backstage objects failed to create secret: Apply faile
d with 1 conflict: conflict with \"manager\" using v1: .metadata.ownerReferences[uid=\"eb0a158e-01a6-4b9e-84d3-8fc61fc43072\"]"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /home/asoro/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:329
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /home/asoro/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /home/asoro/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:227
2024-08-21T11:19:42+02:00       DEBUG   patch object    {"controller": "backstage", "controllerGroup": "rhdh.redhat.com", "controllerKind": "Backstage", "Backstage": {"name":"bs1","namespace"
:"my-ns"}, "namespace": "my-ns", "name": "bs1", "reconcileID": "403a163e-3e39-48cb-8800-ccbcb9406d72", "*v1.ConfigMap": "backstage-appconfig-bs1"}
2024-08-21T11:19:42+02:00       ERROR   Reconciler error        {"controller": "backstage", "controllerGroup": "rhdh.redhat.com", "controllerKind": "Backstage", "Backstage": {"name":"bs1","na
mespace":"my-ns"}, "namespace": "my-ns", "name": "bs1", "reconcileID": "403a163e-3e39-48cb-8800-ccbcb9406d72", "error": "failed to apply backstage objects failed to create secret: Apply faile
d with 1 conflict: conflict with \"manager\" using v1: .metadata.ownerReferences[uid=\"eb0a158e-01a6-4b9e-84d3-8fc61fc43072\"]"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /home/asoro/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:329
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /home/asoro/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /home/asoro/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:227
```

</details>

Could you look into this?

/hold

**Comment by kadel** [comment]
Good catch. I'll look into this.
But the priority of this just becomes lower. This was originally a blocker for implementing PVC support for 1.3, but we won't be adding this to next release, so we can figure out this later.

**Comment by gazarenkov** [comment]
> Thanks for this! I tested it, and while it works great for new installs of the operator, I came across an issue when testing the upgrade path, from 1.2.x to this branch, with an existing operator-backed deployment.
> 
> Steps to reproduce the upgrade path (which can also be reproduced by running `make test-e2e-upgrade [IMG=<image-build-of-the-operator-from-this-branch>]`):
> 
> 1. Switch to the 1.2.x branch and run the operator with `make install run`
> 2. In a separate terminal, create a CR. I used the simple https://github.com/redhat-developer/rhdh-operator/blob/1.2.x/examples/bs1.yaml
> 3. Wait for the application pods to be running and ready:
> 
> oc describe backstage bs1
> 4. Stop the operator running from step 1
> 5. Switch to this PR branch and run the operator with `make install run`
> 
> At this point, it should reconcile successfully, but it looks like there is a reconciliation error:
> 
> ```
> $ oc describe backstages.rhdh.redhat.com/bs1                                   
> Name:         bs1
> Namespace:    my-ns
> Labels:       <none>
> Annotations:  <none>
> API Version:  rhdh.redhat.com/v1alpha2
> Kind:         Backstage
> Metadata:
>   Creation Timestamp:  2024-08-21T09:14:08Z
>   Generation:          1
>   Resource Version:    65153
>   UID:                 eb0a158e-01a6-4b9e-84d3-8fc61fc43072
> Status:
>   Conditions:
>     Last Transition Time:  2024-08-21T09:19:41Z
>     Message:               failed to apply backstage objects failed to create secret: Apply failed with 1 conflict: conflict with "manager" using v1: .metadata.ownerReferences[uid="eb0a158e-01a6-4b9e-84d3-8fc61fc43072"]
>     Reason:                DeployFailed
>     Status:                False
>     Type:                  Deployed
> Events:                    <none>
> ```
> 
> I found a lot of these errors in the operator logs:
> 
> ```
> failed to apply backstage objects failed to create secret: Apply failed with 1 conflict: conflict with \"manager\" using v1: .metadata.ownerReferences[uid=\"eb0a158e-01a6-4b9e-84d3-8fc61fc43072\"]
> ```
> 
> Full logs
> Could you look into this?
> 
> /hold

this is presumably fixed

**Review by rm3l — APPROVED** [review]

**Pull request merged**
